components/suggestion/search_suggestion_tagged_input.jsx

```diff
@@ -0,0 +1,212 @@
+        pretext: PropTypes.string,
+        onKeyDown: PropTypes.func,
+        value: PropTypes.string,
+        onFocus: PropTypes.func,
+    }
+    constructor(props) {
+        super(props);
+        this.state = {
+            value: props.value, pairs: [], remainder: '',
+        };
+    }
+
+    onKeyDown = (e) => {
+        const value = this.getInput().value;
+
+        if (e.key === 'Backspace' && !value) {
+            const pairs = [...this.state.pairs];
+            pairs.pop();
+            this.setState({
+                pairs,
+            }, () => {
+                this.handleChange();
+            });
+        }
+
+        if (this.props.onKeyDown) {
+            this.props.onKeyDown(e);
+        }
+    }
+
+    isTag = (str) => {
+        const lcStr = str.toLowerCase();
+        return exampleTags.some((tag) => lcStr.startsWith(tag));
+    }
+
+    includesTags = (str) => {
+        const lcStr = str.toLowerCase();
+        return exampleTags.some((tag) => lcStr.includes(tag));
+    }
+
+    determinePairs(value = this.props.value) {
+        const includesTags = this.includesTags(value);
+        let remainder;
+        let pairs = [];
+
+        if (includesTags) {
+            const split = value.split(/(?<!:) /);
+            let lastEntry = split[split.length - 1];
+            const lastEntryIsTag = this.isTag(lastEntry);
+
+            lastEntry = lastEntry.split(':');
+
+            remainder = lastEntry[lastEntryIsTag ? 1 : 0];
+            pairs = split.slice(0, split.length - 1).map((x) => {
+                if (this.isTag(x)) {
+                    const [tag, val] = x.split(':');
+                    return {
+                        tag,
+                        value: val,
+                    };
+                }
+
+                return {
+                    tag: '',
+                    value: x,
+                };
+            });
+
+            if (lastEntryIsTag) {
+                pairs.push({tag: lastEntry[0], value: ''});
+            }
+
+            pairs = pairs.filter((pair) => Boolean(pair.tag || pair.value));
+        } else {
+            remainder = value;
+        }
+
+        return {
+            pairs, remainder,
+        };
+    }
+
+    getPretext = () => {
+        const input = this.getInput();
+        let inputValue = '';
+        if (input) {
+            inputValue = input.value.substring(0, input.selectionEnd);
+        }
+
+        return (this.getPairsText() + inputValue).toLowerCase();
+    }
+
+    getPairsText = () => {
+        const {pairs} = this.state;
+        let pairsText = pairs.map(({tag, value = ''}) => {
+            if (tag) {
+                return `${tag}:${value}`;
+            }
+
+            return value;
+        }).join(' ');
+
+        if (pairs.length > 0 && Boolean(pairs[pairs.length - 1].value)) {
+            pairsText += ' ';
+        }
+
+        return pairsText;
+    }
+
+    componentWillReceiveProps(nextProps) {
```

Comment:
This method is deprecated. Can this be done with `getDerivedStateFromProps` instead? I think we'd just have to make `determinePairs` and its helpers static or pull them out of the component entirely for that.

Also, can you make sure this logic only triggers when `nextProps.value !== this.props.value` to avoid unnecessary state changes?